interop/tapir/src/main/scala/caliban/interop/tapir/HttpAdapter.scala

```diff
@@ -0,0 +1,135 @@
+package caliban.interop.tapir
+
+import caliban._
+import caliban.interop.tapir.TapirAdapter._
+import sttp.model.{ headers => _, _ }
+import sttp.tapir.Codec.JsonCodec
+import sttp.tapir.model.ServerRequest
+import sttp.tapir.server.ServerEndpoint
+import sttp.tapir.{ headers, _ }
+import zio._
+
+sealed trait HttpAdapter[-R, E] { self =>
+  protected val endpoints: List[PublicEndpoint[(GraphQLRequest, ServerRequest), TapirResponse, GraphQLResponse[E], Any]]
+
+  protected def executeRequest(
+    graphQLRequest: GraphQLRequest,
+    serverRequest: ServerRequest
+  ): ZIO[R, TapirResponse, GraphQLResponse[E]]
+
+  def serverEndpoints[R1 <: R]: List[ServerEndpoint[Any, RIO[R1, *]]] = {
+    def logic(request: (GraphQLRequest, ServerRequest)): RIO[R1, Either[TapirResponse, GraphQLResponse[E]]] = {
+      val (graphQLRequest, serverRequest) = request
+      executeRequest(graphQLRequest, serverRequest).either
+    }
+    endpoints.map(_.serverLogic(logic))
+  }
+
+  def configure[R1](configurator: ZLayer[R1 & ServerRequest, TapirResponse, R]): HttpAdapter[R1, E] =
```

Comment:
Should I name this one `configureLayer`?

Reply:
Is this layer meant for request interception alone? Perhaps we could name it `withRequestInterception`? If we use the term "configure" might create some confusion that it should be used with the `Configurator` case class somehow

Reply:
@kyri-petrou it should now all be combined for efficiency. So I'd prefer to keep it as `configure` and people can use it with the configurator

Reply:
How about the naming of the 2 variants though? Should I keep the overload or name it `configure`/`configureLayer`?

Reply:
One idea: we could have this:
```scala
type RequestInterceptor[R1, R]: ZLayer[R1 & ServerRequest, TapirResponse, R]
type Configurator[R1]: URIO[R1 & Scope, Unit]

def intercept[R1](interceptor: RequestInterceptor[R1, R]): HttpAdapter[R1, E]
def configure[R1](configurator: Configurator[R1]): HttpAdapter[R & R1, E]
```

Even though under the hood we use the same thing. Since it's now optimized, you can even call them and the layers will be combined.

What do you think?

Reply:
Personally, I really like it. I think it's a good middleground hiding implementation details from the users while retaining efficiency

Reply:
Do we need the `ZIO` variant at all if we have `ZLayer`? My thought for how people would use the `ZIO` configurator variants would only be to roll their own logic there:

```scala
val myConfig = ZLayer.scoped(for {
  request  <- ZIO.service[ServerRequest]
  session  <- Auth.run(request)
  _            <- Configurator.setAllowIntrospection(true).when(session.user.isSuperAdmin)
  _.           <- Configuration.setSkipValidation(true).when(session.isInternal)  
} yield session)
```

But I would think having one way to do things is easier (meaning the `Layer` since it has the most power

Reply:
The problem is that it doesn't work well when all you want is just skip validations. `Configurator. setSkipValidation` returns a layer of `Unit`, which doesn't match the signature of `configure` which expects `R` as a result. So you have to do some annoying conversion which is what the second `configure` allow you to avoid.

Reply:
If I do my suggestion above (intercept/configure), I can delete the methods in Configurator that return a ZLayer and keep only the ones that return ZIO.

The flows would be:
- for simple configuration: use configure + Configurator.set (ZIO)
- for complex workflows: use intercept + ZLayer (you're free to use Configurator in there, or you can cal configure separately)

I think it's pretty good like this!

Reply:
Done!